test/opt/loop_optimizations/hoist_access_chains.cpp

```diff
@@ -0,0 +1,387 @@
+// Copyright (c) 2022 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "gmock/gmock.h"
+#include "source/opt/licm_pass.h"
+#include "test/opt/pass_fixture.h"
+
+namespace spvtools {
+namespace opt {
+namespace {
+
+using ::testing::UnorderedElementsAre;
+using PassClassTest = PassTest<::testing::Test>;
+
+/*
+  Tests for the LICM pass to check it handles access chains correctly
+
+  Generated from the following GLSL fragment shader
+--eliminate-local-multi-store has also been run on the spv binary
+#version 460
+layout(constant_id = 0) const uint sc = 0;
+void main() {
+    for (uint i = 0; i < 2u; ++i) {
+        vec2 invariant_a = vec2(0.0f);
+        float invariant_b = invariant_a.x;
+    }
+    for (uint i = 0; i < 2u; ++i) {
+        vec2 invariant_c = vec2(0.0f);
+        float variant_d = invariant_c[i];
+    }
+    for (uint i = 0; i < 2u; ++i) {
+        vec2 variant_e = vec2(i);
+        float variant_f = variant_e.x;
+    }
+    for (uint i = 0; i < 2u; ++i) {
+        vec2 variant_g = vec2(0.0f);
+        variant_g.x = 1.0f;
+    }
+    for (uint i = 0; i < 2u; ++i) {
+        vec2 invariant_h = vec2(0.0f);
+        float invariant_i = invariant_h[sc];
+    }
+}
```

Comment:
We should talk about what you really want.  All of the test cases involve function scope variables. Many of these could be handled other ways. Ideally all of these loads and stores could be removed by other passes, so the loop end up empty anyway.

You will also need tests to check that you do not move memory operations that should not be moved. For example, if there is shared memory with some type of memory barrier in the loop. Memory operations cannot be moved out of an if statement because that will change the semantics. This applies to the loop condition variable. You cannot move the load or store unless you prove there is a least 1 iteration. Please write many more tests.

Reply:
Thanks for the review @s-perron. The main motivation for my change is to pull out invariant operations, just like `a+b` in the [hoist_simple_case.cpp](https://github.com/KhronosGroup/SPIRV-Tools/blob/main/test/opt/loop_optimizations/hoist_simple_case.cpp). That is a trivial example, but in my use case I have possible invariants located in functions executed many times in a loop, which I want to identify and hoist automatically.

The problem is the previous approach was breaking my shaders (like https://github.com/KhronosGroup/SPIRV-Tools/issues/5075). The additional checks I've added in `ShouldHoistInstruction()` seem to have addressed that, but nothing will be hoisted in that example if `OpStore` isn't considered moveable. But you are right, there are situations I had not considered where moving an `OpStore` or an `OpLoad` could become a problem.

I could simplify this patch so that only motion safe instructions can be moved. What do you think?

Reply:
It's really just adding the `AreInOperandsInvariantInLoop()` check which fixes `LICMPass` if I don't have to consider `OpStore`

Reply:
Yes, simplifying is a good idea.

Reply:
I finally got around to simplifying this fix as explained above, see https://github.com/KhronosGroup/SPIRV-Tools/pull/5087/commits/d31ccad9429cd8a2f1b0a8d4294282148f209795